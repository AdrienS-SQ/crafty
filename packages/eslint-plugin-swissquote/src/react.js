// EcmaScript 6 specific configuration
module.exports = {
  settings: {
    "import/resolver": {
      node: {
        extensions: [".js", ".jsx", ".json"]
      }
    },
    react: {
      version: "15" //Overlaid when loading this file
    }
  },
  parserOptions: {
    ecmaFeatures: {
      jsx: true
    }
  },
  rules: {
    // Recommended rules override
    "@swissquote/swissquote/react/jsx-no-duplicate-props": [
      "error",
      { ignoreCase: true }
    ],
    "@swissquote/swissquote/react/no-deprecated": "warn",
    "@swissquote/swissquote/react/prop-types": [
      "warn",
      { skipUndeclared: true }
    ],

    // Swissquote Rules
    "@swissquote/swissquote/react/jsx-handler-names": "error",
    "@swissquote/swissquote/react/jsx-indent": ["error", 2],
    "@swissquote/swissquote/react/jsx-pascal-case": "error",
    "@swissquote/swissquote/react/no-did-mount-set-state": "error",
    "@swissquote/swissquote/react/no-did-update-set-state": "error",
    "@swissquote/swissquote/react/no-redundant-should-component-update":
      "error",
    "@swissquote/swissquote/react/no-typos": "error",
    "@swissquote/swissquote/react/no-unused-state": "error",
    "@swissquote/swissquote/react/no-will-update-set-state": "error",
    "@swissquote/swissquote/react/prefer-es6-class": ["error", "always"],
    "@swissquote/swissquote/react/prefer-stateless-function": [
      "error",
      { ignorePureComponents: true }
    ],
    "@swissquote/swissquote/react/void-dom-elements-no-children": "error"
  }
};

// Eslint can't load plugins transitively (from a shared config)
// So we have to include the file ourselves and include the rules as if they were ours.
// Solution proposed by @nzakas himself : https://github.com/eslint/eslint/issues/3458#issuecomment-257161846
// replaces `extends: "plugin:react/recommended",`
const reactRules = require("eslint-plugin-react").configs.recommended.rules;
Object.keys(reactRules).forEach(ruleName => {
  // Only define the rules we don't have configured yet
  const key = `@swissquote/swissquote/${ruleName}`
  if (!module.exports.rules.hasOwnProperty(key)) {
    module.exports.rules[key] = reactRules[ruleName];
  }
});
