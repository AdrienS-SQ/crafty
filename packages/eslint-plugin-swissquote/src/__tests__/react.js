/* global describe, it, expect */

const { prepareCLIEngine, lint } = require("../../test_utils");
const engine = prepareCLIEngine(require("../es6"), require("../react"));

function wrap(body) {
  return `
import React from 'react';

export default class MyComponent extends React.Component {
/* eslint no-empty-function: 0, class-methods-use-this: 0 */
${body}
}
`;
}

describe("jsx-no-duplicate-props", () => {
  it("works with different props", () => {
    const result = lint(
      engine,
      `
import * as React from "react";

function SomeComponent() { return <div id="" className="" />;  }
`
    );

    expect(result.messages).toEqual([], "no messages");
    expect(result.warningCount).toBe(0, "no warnings");
    expect(result.errorCount).toBe(0, "no errors");
  });

  it("works fails with the same prop", () => {
    const result = lint(
      engine,
      `
import * as React from "react";

function SomeComponent() { return  <div id="" id="" />;  }
`
    );

    expect(result.messages).toMatchSnapshot();
    expect(result.warningCount).toBe(0, "no warnings");
    expect(result.errorCount).toBe(1, "one error");
  });
});

describe("no-did-mount-set-state", () => {
  it("fails with setState in componentDidMount", () => {
    const result = lint(
      engine,
      `${wrap(`
    componentDidMount() {
      this.setState({
        name: this.props.name.toUpperCase()
      });
    }
    render() {
      return <div>Hello {this.state.name}</div>;
    }
    `)}
    MyComponent.propTypes = {
      name: PropTypes.string.isRequired,
    }`
    );

    expect(result.messages).toMatchSnapshot();
    expect(result.warningCount).toBe(0, "no warnings");
    expect(result.errorCount).toBe(1, "no errors");
  });
});
