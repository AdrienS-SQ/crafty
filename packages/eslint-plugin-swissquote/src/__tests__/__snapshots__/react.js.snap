// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx-no-duplicate-props works fails with the same prop 1`] = `
Array [
  Object {
    "column": 47,
    "endColumn": 52,
    "endLine": 4,
    "line": 4,
    "message": "No duplicate props allowed",
    "nodeType": "JSXAttribute",
    "ruleId": "@swissquote/swissquote/react/jsx-no-duplicate-props",
    "severity": 2,
    "source": "function SomeComponent() { return  <div id=\\"\\" id=\\"\\" />;  }",
  },
]
`;

exports[`no-did-mount-set-state fails with setState in componentDidMount 1`] = `
Array [
  Object {
    "column": 7,
    "endColumn": 20,
    "endLine": 8,
    "line": 8,
    "message": "Do not use setState in componentDidMount",
    "nodeType": "MemberExpression",
    "ruleId": "@swissquote/swissquote/react/no-did-mount-set-state",
    "severity": 2,
    "source": "      this.setState({",
  },
]
`;
